# Jupiter Notebook for Groupwork in AA 2024

#### Import Libraries

In [57]:
import os
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from IPython.display import display
from datetime import timedelta

### 1 Load and Explore Data:

In [58]:
# Load csv-files independently of the operating system (mac or windows or linux) in a pandas dataframe
charging_data = pd.read_csv(os.path.join('Data', 'charging_sessions.csv'))
weather_data = pd.read_csv(os.path.join('Data', 'weather_burbank_airport.csv'))

charging_data = charging_data.rename(columns={'Unnamed: 0': 'counter'})
charging_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   counter           66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.6+ MB


In [59]:
weather_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29244 entries, 0 to 29243
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   city                     29244 non-null  object 
 1   timestamp                29244 non-null  object 
 2   temperature              29219 non-null  float64
 3   cloud_cover              29224 non-null  float64
 4   cloud_cover_description  29224 non-null  object 
 5   pressure                 29236 non-null  float64
 6   windspeed                29158 non-null  float64
 7   precipitation            29244 non-null  float64
 8   felt_temperature         29218 non-null  float64
dtypes: float64(6), object(3)
memory usage: 2.0+ MB


In [60]:
# How do the charging data records look like?
charging_data.head()

counter                        id             connectionTime  \
0        0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1        1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2        2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3        3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4        4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3  2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4  2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...

In [61]:
# How do the weather data records look like?
weather_data.head()

city            timestamp  temperature  cloud_cover  \
0  Burbank  2018-01-01 08:53:00          9.0         33.0   
1  Burbank  2018-01-01 09:53:00          9.0         33.0   
2  Burbank  2018-01-01 10:53:00          9.0         21.0   
3  Burbank  2018-01-01 11:53:00          9.0         29.0   
4  Burbank  2018-01-01 12:53:00          8.0         33.0   

  cloud_cover_description  pressure  windspeed  precipitation  \
0                    Fair    991.75        9.0            0.0   
1                    Fair    992.08        0.0            0.0   
2                    Haze    992.08        0.0            0.0   
3           Partly Cloudy    992.08        0.0            0.0   
4                    Fair    992.08        0.0            0.0   

   felt_temperature  
0               8.0  
1               9.0  
2               9.0  
3               9.0  
4               8.0

In [62]:
# .describe() function only analyses numerical columns
charging_data.describe()

counter  kWhDelivered        siteID        userID
count  66450.000000  66450.000000  66450.000000  49187.000000
mean    4726.102062     11.787916      1.472656   1800.494582
std     3568.025394     10.314789      0.499256   2617.548614
min        0.000000      0.501000      1.000000      1.000000
25%     1845.000000      5.097000      1.000000    431.000000
50%     3960.000000      9.140000      1.000000    818.000000
75%     7027.000000     14.183000      2.000000   1805.000000
max    15291.000000    108.797242      2.000000  19923.000000

In [63]:
# .describe() function only analyses numerical columns
weather_data.describe()

temperature   cloud_cover      pressure     windspeed  precipitation  \
count  29219.000000  29224.000000  29236.000000  29158.000000   29244.000000   
mean      17.885622     30.107993    986.879400      8.566568       0.059360   
std        6.454604      5.400388      3.607065      6.699889       0.471682   
min        2.000000      4.000000    971.000000      0.000000       0.000000   
25%       13.000000     26.000000    984.500000      6.000000       0.000000   
50%       17.000000     33.000000    986.480000      7.000000       0.000000   
75%       22.000000     34.000000    989.110000     13.000000       0.000000   
max       46.000000     47.000000    999.650000     57.000000      18.540000   

       felt_temperature  
count      29218.000000  
mean          17.731433  
std            6.321100  
min            0.000000  
25%           13.000000  
50%           17.000000  
75%           22.000000  
max           42.000000

#### Transform Datatypes
Format the datatypes of the columns to those described in the Team Assignment PDF

In [64]:
# Charging Data
# Ist eine explizite Zuweisung des datatypes "string" notwendig bei columns die schon "object" als datatype haben?
charging_data['connectionTime'] = pd.to_datetime(charging_data['connectionTime'])
charging_data['disconnectTime'] = pd.to_datetime(charging_data['disconnectTime'])
charging_data['doneChargingTime'] = pd.to_datetime(charging_data['doneChargingTime'])
charging_data['siteID'] = charging_data['siteID'].astype(str)
charging_data['userID'] = charging_data['userID'].astype(str)

# Team Assignment PDF: "All datetimes are in UTC (GMT) see timezone field for the correct timezone of the site."
# -> Therefore we have to convert the connectionTime, disconnectTime and doneChargingTime to timezone PST (pacific standard time) as the sites are located in Los Angeles.
charging_data['connectionTime'] = charging_data['connectionTime'].dt.tz_convert(tz = 'America/Los_Angeles')
charging_data['disconnectTime'] = charging_data['disconnectTime'].dt.tz_convert(tz = 'America/Los_Angeles')
charging_data['doneChargingTime'] = charging_data['doneChargingTime'].dt.tz_convert(tz = 'America/Los_Angeles')
charging_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype                              
---  ------            --------------  -----                              
 0   counter           66450 non-null  int64                              
 1   id                66450 non-null  object                             
 2   connectionTime    66450 non-null  datetime64[ns, America/Los_Angeles]
 3   disconnectTime    66450 non-null  datetime64[ns, America/Los_Angeles]
 4   doneChargingTime  62362 non-null  datetime64[ns, America/Los_Angeles]
 5   kWhDelivered      66450 non-null  float64                            
 6   sessionID         66450 non-null  object                             
 7   siteID            66450 non-null  object                             
 8   spaceID           66450 non-null  object                             
 9   stationID         66450 non-null  object                     

In [65]:
# The offset of the columns connectionTime, disconnectTime and doneChargingTime is now -08:00 which is correct for timezone PST (Los Angeles)
charging_data.head()

counter                        id            connectionTime  \
0        0  5e23b149f9af8b5fe4b973cf 2020-01-02 05:08:54-08:00   
1        1  5e23b149f9af8b5fe4b973d0 2020-01-02 05:36:50-08:00   
2        2  5e23b149f9af8b5fe4b973d1 2020-01-02 05:56:35-08:00   
3        3  5e23b149f9af8b5fe4b973d2 2020-01-02 05:59:58-08:00   
4        4  5e23b149f9af8b5fe4b973d3 2020-01-02 06:00:01-08:00   

             disconnectTime          doneChargingTime  kWhDelivered  \
0 2020-01-02 11:11:15-08:00 2020-01-02 09:31:35-08:00        25.016   
1 2020-01-02 14:38:21-08:00 2020-01-02 12:18:05-08:00        33.097   
2 2020-01-02 16:39:22-08:00 2020-01-02 08:35:06-08:00         6.521   
3 2020-01-02 08:38:39-08:00 2020-01-02 07:18:45-08:00         2.355   
4 2020-01-02 14:08:40-08:00 2020-01-02 10:17:30-08:00        13.375   

                                sessionID siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034      1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853      1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993      1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319      1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967      1  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...

In [66]:
# Weather Data
weather_data['timestamp'] = pd.to_datetime(weather_data['timestamp'])
# We assume that the weather data is already in the correct time zone (PST).
weather_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29244 entries, 0 to 29243
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   city                     29244 non-null  object        
 1   timestamp                29244 non-null  datetime64[ns]
 2   temperature              29219 non-null  float64       
 3   cloud_cover              29224 non-null  float64       
 4   cloud_cover_description  29224 non-null  object        
 5   pressure                 29236 non-null  float64       
 6   windspeed                29158 non-null  float64       
 7   precipitation            29244 non-null  float64       
 8   felt_temperature         29218 non-null  float64       
dtypes: datetime64[ns](1), float64(6), object(2)
memory usage: 2.0+ MB


### 2 Check and handle duplicates and missing Data

#### 2.1 Identifying duplicates

In [67]:
charging_data_duplicates = charging_data.duplicated().sum()
weather_data_duplicates = weather_data.duplicated().sum()
# Display the duplicate rows (if any)
print("Duplicate rows:\n charging data -> {}\n weather data -> {}".format(charging_data_duplicates,weather_data_duplicates))


Duplicate rows:
 charging data -> 0
 weather data -> 0


#### 2.2 Identifying missing data

In [75]:
charging_missing_data = charging_data.isnull().sum()
display(charging_missing_data)

weather_missing_data = weather_data.isnull().sum()
display(weather_missing_data[weather_missing_data > 0])
print("Number of unique lines with missing values: " + str(weather_data.isnull().any(axis=1).sum()))

counter                 0
id                      0
connectionTime          0
disconnectTime          0
doneChargingTime     4088
kWhDelivered            0
sessionID               0
siteID                  0
spaceID                 0
stationID               0
timezone                0
userID                  0
userInputs          17263
dtype: int64

temperature                25
cloud_cover                20
cloud_cover_description    20
pressure                    8
windspeed                  86
felt_temperature           26
dtype: int64

Number of unique lines with missing values: 139


#### 2.2 Handling missing data

**Lets look at the missing values**<br>
For `charging_data`:
- `doneChargingTime`: We have 4088 missing values, that is 4088/66450=6,15%. Because we do not want to loose 6% of the data, we will replace the missing values with connectionTime + median charging time.
- `userID`: We have 17263 missing values, that is 17263/66450=25,98%. We will replace them with null values (0.0).
- `userInput`: We have 17263 missing values, that is 17263/66450=25,98%. We will replace them with null values (Null).

For `weather_data`:
- There are 139 lines with missing values through the data set. Because this is only 0,47%, we decided to drop these records. 

In [69]:
## charging data
# Calculate median charging time in seconds and round it
median_charging_time = (charging_data['doneChargingTime'] - charging_data['connectionTime']).dt.total_seconds().median()
median_charging_time = round(median_charging_time)  # Round the median charging time
display(median_charging_time)

# Make a copy to avoid modifying the original DataFrame
charging_data_cleaned = charging_data.copy()

# Fill 'doneChargingTime' NaNs with 'connectionTime' + median charging time
charging_data_cleaned['doneChargingTime'] = charging_data_cleaned['doneChargingTime'].fillna(
    charging_data_cleaned['connectionTime'] + pd.to_timedelta(median_charging_time, unit='s')
)

# Fill userID and userInput with null values -> Change if type changed
charging_data_cleaned['userID'] = charging_data_cleaned['userID'].fillna(0)
charging_data_cleaned['userInputs'] = charging_data_cleaned['userInputs'].fillna("")

## weather data
# Make a copy to avoid modifying the original DataFrame
weather_data_cleaned = weather_data.copy()

# Drop lines with missing values
weather_data_cleaned = weather_data_cleaned.dropna()

10022

### 3 Erroneous Data
### ToDo User Input Data
Format UserInputData into an array?

In [70]:
#ToDo -> Fynn, pls use cleaned df

### ToDo Data correction: Disconnect time is before done charging time
The data shows that is some cases Disconnect time is before done charging time...thats an error...Solution If(DoneTime after DiscTime): DoneTime == DiscTime

In [71]:
"""
tmp_check_doneChargingTime = charging_data[
    charging_data['doneChargingTime'].isnull() & (charging_data['doneChargingTime_clean'] > charging_data['disconnectTime'])]
print(tmp_check_doneChargingTime[['doneChargingTime', 'doneChargingTime_clean', 'disconnectTime']])
# TODO: entpsrechende Einträge korrigieren oder in Zeile 15 schon abändern... oder vielleicht erst unter erroneous Data behandeln, um die anderen Einträge noch mitzunehmen?
"""

# Check whether disconnectTime is before connectionTime
erroneous_disconnect_times = charging_data[charging_data['disconnectTime'] < charging_data['connectionTime']]

# Output the wrong lines
print(erroneous_disconnect_times[['connectionTime', 'disconnectTime']])


Empty DataFrame
Columns: [connectionTime, disconnectTime]
Index: []


In [72]:
# Calculate the loading time in seconds
charging_data['charging_duration'] = (charging_data['doneChargingTime_clean'] - charging_data['connectionTime']).dt.total_seconds()

# Calculate the total connection time in seconds
charging_data['connection_duration'] = (charging_data['disconnectTime'] - charging_data['connectionTime']).dt.total_seconds()

# Check whether the loading time is longer than the connection time
erroneous_charging_times = charging_data[charging_data['charging_duration'] > charging_data['connection_duration']]

# Count the number of rows with erroneous charging times
erroneous_count = erroneous_charging_times.shape[0]

# Display the count
print(f"Number of rows with longer charging duration then connection: {erroneous_count}")

print(erroneous_charging_times[['connectionTime', 'disconnectTime', 'doneChargingTime_clean', 'charging_duration', 'connection_duration']])

KeyError: 'doneChargingTime_clean'

In [182]:
#Identify the differance between the 6436 changed data and the 6441 equal datas one underneath
# Identify lines where doneChargingTime_clean was already equal to disconnectTime
pre_existing_equal_times = charging_data[
    charging_data['doneChargingTime_clean'] == charging_data['disconnectTime']
]

# Count how many of them were marked as faulty
pre_existing_errors = pre_existing_equal_times[
    pre_existing_equal_times['charging_duration'] > pre_existing_equal_times['connection_duration']
]

print(f"Number of rows where doneChargingTime_clean was already equal to disconnectTime: {pre_existing_equal_times.shape[0]}")
print(f"Number of these rows identified as erroneous: {pre_existing_errors.shape[0]}")

Number of rows where doneChargingTime_clean was already equal to disconnectTime: 5
Number of these rows identified as erroneous: 0


In [183]:
#Adjust doneChargingTime_clean if it exceeds disconnectTime
# Set doneChargingTime_clean to disconnectTime where it exceeds disconnectTime
charging_data.loc[charging_data['doneChargingTime_clean'] > charging_data['disconnectTime'], 'doneChargingTime_clean'] = charging_data['disconnectTime']

# Verify adjustments
adjusted_charging_times = charging_data[charging_data['doneChargingTime_clean'] == charging_data['disconnectTime']]
print(f"Number of rows where doneChargingTime_clean was adjusted to disconnectTime: {adjusted_charging_times.shape[0]}")

print(adjusted_charging_times[['connectionTime', 'disconnectTime', 'doneChargingTime_clean']].head())

Number of rows where doneChargingTime_clean was adjusted to disconnectTime: 6441
              connectionTime            disconnectTime  \
12 2020-01-02 15:04:38+00:00 2020-01-02 22:08:39+00:00   
20 2020-01-02 15:28:47+00:00 2020-01-02 19:01:54+00:00   
25 2020-01-02 15:42:05+00:00 2020-01-02 21:58:45+00:00   
26 2020-01-02 15:57:24+00:00 2020-01-02 16:35:37+00:00   
33 2020-01-02 16:34:35+00:00 2020-01-02 18:49:41+00:00   

      doneChargingTime_clean  
12 2020-01-02 22:08:39+00:00  
20 2020-01-02 19:01:54+00:00  
25 2020-01-02 21:58:45+00:00  
26 2020-01-02 16:35:37+00:00  
33 2020-01-02 18:49:41+00:00  


### 4 Outliers

In [ ]:
# Unrealistic Time Differences
# Check if the charging duration is either too short (less than 1 minute) or too long (more than 1 day).

unrealistic_times = charging_data[(charging_data['charging_duration'] < 60) | (charging_data['charging_duration'] > 86400)]

# Show the number of rows with unrealistic charging durations
print(f"Number of rows with unrealistic charging times: {unrealistic_times.shape[0]}")

# Show examples of the data
print("Examples of erroneous charging durations (too short (1 Min) or too long(24 hours)):")
print(unrealistic_times[['connectionTime', 'disconnectTime', 'doneChargingTime_clean', 'charging_duration']].head())


# ToDo: How to handle outliers? How to set the boundries? Is it necassary to handle the outliers (bzw. how much)? -Justin, pls use cleaned df

Number of rows with unrealistic charging times: 95
Examples of erroneous charging durations (too short (1 Min) or too long(24 hours)):
                 connectionTime            disconnectTime  \
18604 2021-07-16 15:45:30+00:00 2021-07-19 23:06:05+00:00   
18606 2021-07-16 16:23:01+00:00 2021-07-19 23:05:42+00:00   
18607 2021-07-16 16:40:11+00:00 2021-07-19 23:05:30+00:00   
18609 2021-07-16 17:52:18+00:00 2021-07-19 20:41:13+00:00   
20425 2019-01-10 01:51:20+00:00 2019-01-11 07:17:28+00:00   

         doneChargingTime_clean  charging_duration  
18604 2021-07-19 20:36:25+00:00           276655.0  
18606 2021-07-19 20:45:36+00:00           274955.0  
18607 2021-07-19 20:45:36+00:00           273925.0  
18609 2021-07-19 20:41:07+00:00           269329.0  
20425 2019-01-11 07:17:28+00:00           106027.0  



### 5 Inconsistancies
Correct missspelling and false dataTypes

Data is conistant after visual check. Final check can be done after data types are chaned...

In [ ]:

#ToDo: Check missspelling etc., dataTypes have been done above -> TBD with final check...

### 5 Feature engineering
Add aditional features (e.g. charging or idle time) -> To be discussed


### 6 Final review
Look again for missing, errorous, and inconsistant data

In [38]:
# Missing data
print("Number of missing values in cleand charging data: " + str(charging_data_cleaned.isnull().sum().sum()))
print("Number of missing values in cleand weather data: " + str(weather_data_cleaned.isnull().sum().sum()))

Number of missing values in cleand charging data: 0
Number of missing values in cleand weather data: 0
